pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/SomaticResultUploadService.java

```diff
@@ -0,0 +1,186 @@
+
+        Map<String, String> headers = Map.of("Content-Type", somaticResultMetaData.getMimeType());
+
+        URL signedURL = storageClient.generateSignedUrl(
+                signer, uploadBucket, blobPath,
+                maxSignedUrlMins, TimeUnit.MINUTES,
+                HttpMethod.PUT, headers);
+
+        return new AuthorizeResult(AuthorizeResultType.OK, signedURL, createdUpload, somaticUploadSettings);
+    }
+
+    public SomaticResultUpload deleteUpload(long userId, int documentId) {
+        SomaticResultUpload deletedSomaticResultUpload = SomaticResultUpload.deleteDocumentByDocumentId(userId, documentId);
+        Blob blobToDelete = storageClient.getBlob(deletedSomaticResultUpload.getBucket(), deletedSomaticResultUpload.getBlobPath());
+        if (blobToDelete != null) {
+            boolean deleted = blobToDelete.delete();
+            if (deleted) {
+                log.info("User {} deleted somatic document {}", userId, documentId);
+            } else {
+                log.error("Somatic document failed to delete from GCS. Manual intervention required.  "
+                                + "Last recorded bucket: {}, blobPath: {} ",
+                        deletedSomaticResultUpload.getBucket(), deletedSomaticResultUpload.getBlobPath());
+                throw(new RuntimeException("Deletion failed, contact DSM developer."));
+            }
+        } else {
+            log.error("Somatic document blob not found in bucket when attempting to delete from GCS. Manual intervention required.  "
+                            + "Last recorded bucket: {}, blobPath: {} ",
+                    deletedSomaticResultUpload.getBucket(), deletedSomaticResultUpload.getBlobPath());
+            throw(new RuntimeException("Deletion failed, contact DSM developer."));
+        }
+        return deletedSomaticResultUpload;
+    }
+
+    public List<SomaticResultUpload> getSomaticResultsForParticipant(String realm, String ddpParticipantId) {
+        return SomaticResultUpload.getSomaticFileUploadDocuments(realm, ddpParticipantId);
+    }
+
+    private static String makeBlobPath(SomaticResultMetaData payload, String userGuid, String studyGuid,
+                                String fileGuid) {
+        return String.format("%s/%s_%s_%s_%s",
+                studyGuid, fileGuid, userGuid, getCurrentTimestamp(), payload.getFileName());
+    }
+
+    private static String getCurrentTimestamp() {
+        return new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss").format(new Date());
+    }
+
+    private String getUploadsBucket(String realm) {
+        if (this.realmToUploadBucketMap.isEmpty()) {
+            throw new RuntimeException("Uploading documents not fully configured.  Please contact a DSM developer");
+        }
+        return this.realmToUploadBucketMap.get(realm.toLowerCase());
+    }
+
+    private enum AuthorizeResultType {
```

Comment:
Seems like some of these would be 400 errors?

Reply:
Maybe.
IMHO, the following errors are worthy of 500s.
138 - we marked the database as deleted but the GCS object wasn't.  There's nothing for a user to do - and I would expect this to alert us.
165 - DSM isn't properly configured for uploads - nothing a user can do.

The exception thrown at 144 might be worthy of a different exception and error handling - but it also might be an internal error (e.g. misconfiguration).  I assumed misconfiguration but can see where we might allow that one to proceed without error.  Have any strong feelings about this before I refactor?

Reply:
I was mostly looking at the AuthorizeResultTypes, which are used earlier in the file.

Reply:
Ah, those are handed back to the client.  It's an existing pattern DSS uses to communicate with their FileUpload pattern there.

Reply:
Discussed in person.